### PR TITLE
Api to allow loading WriteableBitmap from File / Stream

### PR DESCRIPTION
**Pull request by danwalmsley**
## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

**Review by Szpoti — APPROVED** [review]
Code looks good, its a usefull feature to have.
One thing I dont quite understand is why are there a bunch of functions with
``` throw new NotImplementedException(); ``` inside? Will these be changed or they have some other purpose?

File: src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs [hunk 1]
```diff
@@ -15,7 +17,70 @@ internal class WriteableBitmapImpl : IWriteableBitmapImpl, IDrawableBitmapImpl
         private static readonly SKBitmapReleaseDelegate s_releaseDelegate = ReleaseProc;
         private readonly SKBitmap _bitmap;
         private readonly object _lock = new object();
+        
+        /// <summary>
+        /// Create immutable bitmap from given stream.
```
Comment (grokys) [inline src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs:22]: An immutable writeable bitmap?

**Comment by danwalmsley** [comment]
> Code looks good, its a usefull feature to have.
> One thing I dont quite understand is why are there a bunch of functions with
> `throw new NotImplementedException();` inside? Will these be changed or they have some other purpose?

Ah these are because they are test stubs, and these methods dont get tested from every test stub (or even at all)

hence notimplemented... but those methods are not part of the actual framework, just the testing part.

File: src/Avalonia.Visuals/Platform/IPlatformRenderInterface.cs [hunk 2]
```diff
@@ -100,6 +100,38 @@ public interface IPlatformRenderInterface
...
+        
+        /// <summary>
+        /// Loads a WriteableBitmap implementation from a file..
+        /// </summary>
+        /// <param name="fileName">The filename of the bitmap.</param>        
+        /// <returns>An <see cref="IBitmapImpl"/>.</returns>
+        IWriteableBitmapImpl LoadWriteableBitmap(string fileName);
+
+        /// <summary>
+        /// Loads a WriteableBitmap implementation from a file..
```
Comment (MarchingCube) [inline src/Avalonia.Visuals/Platform/IPlatformRenderInterface.cs:129]:
```suggestion
        /// Loads a WriteableBitmap implementation from a file.
```

File: src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs [hunk 3]
```diff
@@ -15,7 +17,70 @@ internal class WriteableBitmapImpl : IWriteableBitmapImpl, IDrawableBitmapImpl
...
+            {
+                _bitmap = SKBitmap.Decode(skiaStream);
+
+                if (_bitmap == null)
+                {
+                    throw new ArgumentException("Unable to load bitmap from provided data");
+                }
+
+                PixelSize = new PixelSize(_bitmap.Width, _bitmap.Height);
+                Dpi = new Vector(96, 96);
```
Comment (MarchingCube) [inline src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs:37]:
```suggestion
                Dpi = SkiaPlatform.DefaultDpi;
```

File: src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs [hunk 4]
```diff
@@ -15,7 +17,70 @@ internal class WriteableBitmapImpl : IWriteableBitmapImpl, IDrawableBitmapImpl
...
+                {
+                    var scaledBmp = bmp.Resize(desired, interpolationMode.ToSKFilterQuality());
+                    bmp.Dispose();
+                    bmp = scaledBmp;
+                }
+
+                _bitmap = bmp;
+
+                PixelSize = new PixelSize(bmp.Width, bmp.Height);
+                Dpi = new Vector(96, 96);
```
Comment (MarchingCube) [inline src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs:80]:
```suggestion
                Dpi = SkiaPlatform.DefaultDpi;
```

**Review by MarchingCube — CHANGES_REQUESTED** [review]
Mostly fine but there are a few smaller problems.

File: src/Avalonia.Visuals/Platform/IPlatformRenderInterface.cs (hunk 2 above)
Comment (danwalmsley) [reply src/Avalonia.Visuals/Platform/IPlatformRenderInterface.cs]: doine.

File: src/Avalonia.Visuals/Platform/IPlatformRenderInterface.cs [hunk 5]
```diff
@@ -100,6 +100,38 @@ public interface IPlatformRenderInterface
...
+        /// Loads a WriteableBitmap implementation from a stream to a specified height maintaining aspect ratio.
+        /// </summary>
+        /// <param name="stream">The stream to read the bitmap from.</param> 
+        /// <param name="height">The desired height of the resulting bitmap.</param>
+        /// <param name="interpolationMode">The <see cref="BitmapInterpolationMode"/> to use should resizing be required.</param>
+        /// <returns>An <see cref="IBitmapImpl"/>.</returns>
+        IWriteableBitmapImpl LoadWriteableBitmapToHeight(Stream stream, int height, BitmapInterpolationMode interpolationMode = BitmapInterpolationMode.HighQuality);
+        
+        /// <summary>
+        /// Loads a WriteableBitmap implementation from a file..
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Visuals/Platform/IPlatformRenderInterface.cs]: done.

File: src/Avalonia.Visuals/Media/Imaging/WriteableBitmap.cs [hunk 6]
```diff
@@ -34,8 +37,50 @@ public WriteableBitmap(PixelSize size, Vector dpi, PixelFormat format, AlphaForm
...
+        }
+
+        /// <summary>
+        /// Loads a Bitmap from a stream and decodes at the desired height. Aspect ratio is maintained.
+        /// This is more efficient than loading and then resizing.
+        /// </summary>
+        /// <param name="stream">The stream to read the bitmap from. This can be any supported image format.</param>
+        /// <param name="height">The desired height of the resulting bitmap.</param>
+        /// <param name="interpolationMode">The <see cref="BitmapInterpolationMode"/> to use should any scaling be required.</param>
+        /// <returns>An instance of the <see cref="Bitmap"/> class.</returns>
```
Comment (MarchingCube) [inline src/Avalonia.Visuals/Media/Imaging/WriteableBitmap.cs:76]:
```suggestion
        /// <returns>An instance of the <see cref="WriteableBitmap"/> class.</returns>
```

**Review by MarchingCube — APPROVED** [review]
LGTM. Noticed two minor doc issues that I've missed before.

**Pull request merged**